### ISSUE TITLE
Feature request: Require explicit array creation before inserting values

### ISSUE THREAD

**Issue by Levivb** [slevomat/coding-standard#418]
Hey,

For a while now (don't know since when) php allows the implicit creation of arrays like so:
```php
function(){
    $array[] = 'value-1';
    $array[] = 'value-2';
    return $array;
}
```
Which will result in:
```php
[
     "value-1",
     "value-2",
]
```

This implicit creation of arrays is, however, bad practice and can result in bugs which are hard to pinpoint.

For instance, I once refactored to the following (simplified):
```php
$attr = $this->params();
$applications = $this->getApplications();
foreach($applications as $application) {
    $attr['now'] = Carbon\Carbon::now();
    $application->fill($attr);
    $application->save();
}
```

To this:
```php
$attr = $this->params();
$applications = $this->getApplications();
$applications->each(function(Application $application) {
    $attr['now'] = Carbon\Carbon::now();
    $application->fill($attr);
    $application->save();
});
```

Note how `$attr` in the closure is not passed to the `use` section of the anonymous function.

Although `$attr` is not defined in the function, due to php's forgiving nature, the `$attr` is automatically created upon inserting a value. But this time, with `now` as the only value.

The refactor itself took me 5 minutes, the havoc it caused took me 40 hours of scrolling through logs, uploaded files and recovering lost form-data.

My request: I would like to see a sniff which prohibits the implicit creation of arrays.
_It seems there are more people running into the same issue: https://stackoverflow.com/a/47674196_

Let me know what you think.

Thanks in advance!

**Comment by kukulich** [comment]
I think it's interesting idea :)

**Comment by kukulich** [comment]
Implemented in https://github.com/slevomat/coding-standard/commit/d1342975c29cd465f5d4e2bf3e736e138980a16e
